sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def inputTypes: Seq[DataType] = {
+    // TODO: properly handle String Types with differing collations
+    val leftIsString = left.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    val rightIsString = right.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
+
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
```

Comment:
I think that you should just fail if collations are not equal at this point. Maybe you can create some common trait that will assure that collations are equal (e.g. in `StringPredicate`).

Reply:
That would currently disable queries like:

- `contains("abc", collate("C", "UNICODE_CI"))` // 0, 3
- `contains(collate("abc", "UNICODE_CI"), "C")` // 3, 0

And only leave enabled queries like:

- `contains(collate("abc", "UNICODE_CI"), collate("C", "UNICODE_CI"))` // 3, 3
- `contains("abc", "c")` // 0, 0

Shouldn't we essentially allow different collationIDs, as long as they are "compatible"? (whatever that might mean in the context of the current collatorFactory implementation)

Reply:
We will take care of literal casting as a separate track. @stefankandic as FYI.
But I don't think that every operator needs to do such checks.